File: src/algorithms/subproblem.jl
using LinearAlgebra
using Roots

function subproblem(hess::AbstractMatrix{T}, grad::Vector{T}, delta::T; tol::T = 1e-4, max_iters::Int = 100, verbose::Int = 2) where T <: Float64

	n 					= length(grad)
	sol 				= zeros(T, n)
	regularized_hess 	= similar(hess)
	lambda 				= zero(T)

	if !LinearAlgebra.isposdef(hess)
		eigen_sys = eigen(hess)
		eigen_val_min = minimum(real(eigen_sys.values))
		lambda = abs(eigen_val_min) + 0.00001
		if verbose > 1
			println("not positive definite: eigenvalue min = ", eigen_val_min)
		end
	end

	for i in 1:max_iters

		# Regularize Hessian if needed
		regularized_hess .= 0.5(hess' + hess) + lambda * I

		# Newton step
		#RTR = regularized_hess_ch' * regularized_hess_ch
		sol .= regularized_hess \ -grad
		println("newton step = ", sol);
		println("Lambda = ", lambda)
		sol_norm = norm(sol)
		println("sol_norm = ", sol_norm)

		if sol_norm ≤ delta
			if lambda == 0 || abs(sol_norm - delta ) < tol * delta
				println("Newton step finished")
				if verbose > 0
					println("Newton step ")
				end
				return sol
			else
				eigen_vec_min = (eigen(hess, sortby=minimum).vectors)[:,1]
				fn = x -> sqrt((sol + x * eigen_vec_min)' * (sol + x * eigen_vec_min)) - delta
				root = find_zero(fn, delta/100.)
				if verbose > 1
					prinln("hard case solution")
				end
				return sol + root * eigen_vec_min
			end
		else
			ql_norm = norm((LinearAlgebra.cholesky(regularized_hess).U)' \ sol)
			lambda += (sol_norm/ql_norm)^2 * ((sol_norm - delta)/delta);
			#println("inner error = ", sol_norm - delta )
		end

		# if abs(sol_norm - delta ) < tol * delta
		# 	if (verbose > 0)
		# 		println("sub problem solution was found after i = ", i , " iterations", "  sol_nomr = ", sol_norm)
		# 	end
		# 	break
		# end

		if i == max_iters
			println("no solution in inner loop !!!!")
		end

	end
	return sol
end





# function subproblem(hess::AbstractMatrix{T}, grad::Vector{T}, delta::T; tol::T = 1e-4, max_iters::Int = 100) where T <: Real
#     # Initialization
#     n = length(grad)
#     sol = zeros(T, n)
#     λ = zero(T)

#     # Ensure Hessian is positive definite
#     if !isposdef(hess)
#         eigen_sys = eigen(hess)
#         eigen_val_min = minimum(real(eigen_sys.values))
#         λ = abs(eigen_val_min) + 1e-4  # Regularization factor
#     end

#     # Iterative refinement
#     for iter in 1:max_iters
#         # Regularize Hessian
#         regularized_hess = hess + λ * I

#         # Solve Newton step
#         sol_newton = regularized_hess \ -grad
#         sol_norm = norm(sol_newton)

#         if sol_norm ≤ delta
#             # Check if the Newton step is valid or needs refinement
#             if λ == 0 || abs(sol_norm - delta) < tol * delta
#                 return sol_newton
#             else
#                 # Hard case: Find root to project onto the boundary
#                 eigen_sys = eigen(hess, sortby=minimum)
#                 eigen_vec_min = eigen_sys.vectors[:, 1]
#                 fn = x -> norm(sol_newton + x * eigen_vec_min) - delta
#                 root = find_zero(fn, delta / 100)
#                 return sol_newton + root * eigen_vec_min
#             end
#         else
#             # Update λ to refine the regularization
#             ql_norm = norm(cholesky(regularized_hess).U' \ sol_newton)
#             λ += (sol_norm / ql_norm)^2 * ((sol_norm - delta) / delta)
#         end

#         # Convergence check
#         if abs(sol_norm - delta) < tol * delta
#             break
#         end

#         if iter == max_iters
#             println("Warning: Subproblem did not converge within max iterations!")
#         end
#     end

#     return sol
# end

File: src/algorithms/subproblem.c

#include "utilities.h"


// subproblem function
void	subproblem(const T hess[N][N], const T grad[N], T sol[N], T *delta) {
	T regularized_hess[N][N];
	T chol[N][N];
	T temp[N];
	T lambda = 0.0;
	T smallest_eigenvector[N];


	//Initialize solution to zero
	for (int i = 0; i < N; i++) {
		sol[i] = 0.0;
	}

	// Check if the Hessian is positive definite
	if (!is_positive_definite(hess)) {
		// Estimate minimum eigenvalue (simplified for performance)
		T eigen_min = find_smallest_eigenvalue(hess, smallest_eigenvector);
		lambda = fabs(eigen_min) + 0.0001;
	}

	// Main loop
	for (int iter = 0; iter < subproblem_max_iters; iter++) {
		// Regularize the Hessian: regularized_hess = 0.5 * (hess + hess^T) + lambda * I
		for (int i = 0; i < N; i++) {
			for (int j = 0; j < N; j++) {
				regularized_hess[i][j] = 0.5 * (hess[i][j] + hess[j][i]);
			}
			regularized_hess[i][i] += lambda;
		}

		// Perform Cholesky decomposition: chol * chol^T = regularized_hess
		cholesky_decomposition(regularized_hess, chol);

		// Solve the Newton step: sol = regularized_hess \ -grad
		for (int i = 0; i < N; i++) {
			temp[i] = -grad[i];
		}

		// Forward substitution: solve chol * y = -grad
		T y[N];
		forward_substitution(chol, temp, y);

		// Backward substitution: solve chol^T * x = y
		backward_substitution(chol, y, sol);

		// Check norm of the solution
		T sol_norm = vector_norm(sol);
		if (sol_norm <= *delta) {
			if (lambda == 0 || fabs(sol_norm - *delta) < subproblem_tol * *delta) {
				return;
			} else {
				lambda = find_root(sol, smallest_eigenvector, *delta, 0.0);
				// Hard case handling can be added here if needed
				return;
			}
		} else {
			// Update lambda
			T ql_norm = vector_norm(temp);  // Use forward_substitution result as an approximation
			lambda += pow(sol_norm / ql_norm, 2) * ((sol_norm - *delta) / *delta);
		}
	}
}
